syntheseus/search/algorithms/pdvn.py

```diff
@@ -0,0 +1,372 @@
+        costs = self.and_node_cost_fn(and_nodes, graph=graph)
+        assert len(costs) == len(and_nodes)
+        for node, cost in zip(and_nodes, costs):
+            node.data["pdvn_reaction_cost"] = cost
+
+    def choose_successors_to_visit(
+        self,
+        node: ANDOR_NODE,  # type: ignore[override]  # AND/OR vs any graph node
+        graph: AndOrGraph,
+    ) -> Sequence[ANDOR_NODE]:
+        """
+        This function performs the "select" step of MCTS. It chooses which child node(s) to visit.
+
+        In normal MCTS a single child is always chosen using an upper confidence bound. In PDVN-MCTS,
+        the same logic is used to select a child at an OrNode (selecting a single reaction). However,
+        at an AndNode the logic is different: a single OrNode child is chosen using the logic in
+        section 3.2 of Liu et al 2023, which states:
+
+            To select a child molecule node, we prioritize molecules that have not been expanded;
+            if none are available, we choose ones that have not been solved. If molecule nodes are
+            either all expanded or all solved, we randomly select one of them.
+
+        The logic below follows the paragraph above, disregarding that the final clause
+        will only be reached if the childre are all expanded *and* all solved, rather than *or*
+        as it says in the text.
+        """
+        if isinstance(node, OrNode):
+            # This is equivalent to normal MCTS. Choose single child based on upper confidence bound
+            output = super().choose_successors_to_visit(node, graph)
+        elif isinstance(node, AndNode):
+            children = list(graph.successors(node))
+            unsolved_children = [c for c in children if not c.has_solution]
+            expandable_children = [c for c in children if self.can_expand_node(c, graph)]
+
+            # Choose a child at random from either unexpanded, unsolved, or all children
+            if len(expandable_children) > 0:
+                # First priority: expand an unexpanded child, if it exists
+                children_to_choose_from = expandable_children
+            elif len(unsolved_children) > 0:
+                # Second priority: expand an unsolved node, if it exists
+                children_to_choose_from = unsolved_children
+            else:
+                # Third priority: choose a random node
+                children_to_choose_from = children
+            output = [self.random_state.choice(children_to_choose_from)]
+        else:
+            raise TypeError(f"Unexpected node type: {type(node)}")
+
+        # Ensure only 1 node was chosen
+        assert len(output) == 1, "PDVN_MCTS can only choose 1 successor to visit."
+        return output  # type: ignore[return-value]  # mypy thinks it could be any graph node due to super() call
+
+    def _get_leaf_node_reward(self, node: OrNode, graph: AndOrGraph) -> tuple[float, float]:  # type: ignore[override]
```

Comment:
This is fine for now, but eventually should we make the reward type generic @AustinT?

Reply:
Yes I agree. I considered doing it with this PR but didn't want to introduce a lot of additional complexity for what might end up being a "one off" algorithm.